### PR TITLE
fix: Added Validation for Insurance input fields in patient registration  issue (#8672)

### LINKED ISSUE #8672: Add Validation for Insurance input fields in patient registration form
**Describe the bug**

Currently, the patient registration form lacks frontend validation for the insurance fields, and the backend only validates the member ID and policy ID.

**To Reproduce**

1. Go to the patient tab
2. Click on add patients
3. scroll down and click Add insurance
4. without filling in the fields submit it
5. See an error

**Expected behavior**

- [ ] add backend validation for insurer name 
- [ ] add frontend validation for member id, policy id and insurer name

**Screenshots**

image

### PR DESCRIPTION
**Pull request by syedfardeenjeelani**
## Proposed Changes

- Fixes #8672
- Added frontend validation for insurance input fields (Member ID, Policy ID, and Insurer Name) in the patient registration form.

https://github.com/user-attachments/assets/2da86816-e638-44c3-b756-7d015e7ddf20

@ohcnetwork/care-fe-code-reviewers

## Merge Checklist

- [x] Prep screenshot or demo video for changelog entry, and attach it to the issue.

### CODE REVIEW

**Comment by rithviknishad** [comment]
@syedfardeenjeelani is this ready for review?

**Comment by syedfardeenjeelani** [comment]
> @syedfardeenjeelani is this ready for review?

 we are getting the error from the validator.ts which is something like this
if (
!value.policy_id.trim() ||
!value.subscriber_id.trim() ||
(enable_hcx && (!value.insurer_id?.trim() || !value.insurer_name?.trim()))
)
return "All fields are mandatory";

so should i use field_required from the i18n locale here instead of "All fields are mandatory"

**Comment by rithviknishad** [comment]
yes

**Comment by syedfardeenjeelani** [comment]
> yes

Thanks I'll do the changes and push

**Comment by syedfardeenjeelani** [comment]
@rithviknishad  if we do it like this 
 if (
    !value.policy_id.trim() ||
    !value.subscriber_id.trim() ||
    (enable_hcx && (!value.insurer_id?.trim() || !value.insurer_name?.trim()))
  )
    return t("field_required");
};


then it will also show it right here
issue
  so instead should a make a key which  says  "all_fields_mandatory": "All fields are mandatory"

or should we handle both cases seprately

**Comment by rithviknishad** [comment]
Let's keep the error texts same as before itself instead of showing the error on ALL fields, just that ensure ALL fields required validation happens as mentioned in the issue.

**Comment by syedfardeenjeelani** [comment]
@rithviknishad  I've made some changes based on your feedback, but I'm not sure if I got it right. Could you take a look at my PR when you get a chance and let me know if i am headed in the right direction or not ?

File: src/Components/HCX/InsuranceDetailsBuilder.tsx [hunk 1]
```diff
@@ -163,7 +169,9 @@ const InsuranceDetailEditCard = ({
             />
             <TextFormField
               name="insurer_name"
+              required
               label={t("policy__insurer_name")}
+              error={error}
```
Comment (rithviknishad) [inline src/Components/HCX/InsuranceDetailsBuilder.tsx]:
```suggestion
```

File: src/Components/HCX/InsuranceDetailsBuilder.tsx [hunk 2]
```diff
@@ -133,6 +138,7 @@ const InsuranceDetailEditCard = ({
         <TextFormField
           required
           name="policy_id"
+          error={error}
```
Comment (rithviknishad) [inline src/Components/HCX/InsuranceDetailsBuilder.tsx]:
```suggestion
```

File: src/Components/HCX/InsuranceDetailsBuilder.tsx [hunk 3]
```diff
@@ -125,6 +129,7 @@ const InsuranceDetailEditCard = ({
         <TextFormField
           required
           name="subscriber_id"
+          error={error}
```
Comment (rithviknishad) [inline src/Components/HCX/InsuranceDetailsBuilder.tsx]:
```suggestion
```

File: src/Components/HCX/InsuranceDetailsBuilder.tsx [hunk 4]
```diff
@@ -91,12 +93,14 @@ const InsuranceDetailEditCard = ({
...
   handleRemove,
   gridView,
+  error,
 }: {
   policy: HCXPolicyModel;
   handleUpdate: (event: FieldChangeEvent<unknown>) => void;
   handleUpdates: (diffs: object) => void;
   handleRemove: () => void;
   gridView?: boolean;
+  error: FieldError;
```
Comment (rithviknishad) [inline src/Components/HCX/InsuranceDetailsBuilder.tsx]:
```suggestion
```

File: src/Components/HCX/InsuranceDetailsBuilder.tsx [hunk 5]
```diff
@@ -91,12 +93,14 @@ const InsuranceDetailEditCard = ({
   handleUpdates,
   handleRemove,
   gridView,
+  error,
```
Comment (rithviknishad) [inline src/Components/HCX/InsuranceDetailsBuilder.tsx]:
```suggestion
```

File: src/Components/HCX/InsuranceDetailsBuilder.tsx [hunk 6]
```diff
@@ -73,6 +74,7 @@ export default function InsuranceDetailsBuilder(props: Props) {
           >
             <InsuranceDetailEditCard
               policy={policy}
+              error={props.error}
```
Comment (rithviknishad) [inline src/Components/HCX/InsuranceDetailsBuilder.tsx]:
```suggestion
```

File: src/Components/HCX/InsuranceDetailsBuilder.tsx [hunk 7]
```diff
@@ -15,6 +15,7 @@ import request from "../../Utils/request/request";
 import routes from "../../Redux/api";
 import { useTranslation } from "react-i18next";
 import careConfig from "@careConfig";
+import { FieldError } from "../Form/FieldValidators";
```
Comment (rithviknishad) [inline src/Components/HCX/InsuranceDetailsBuilder.tsx]:
```suggestion
```

File: src/Redux/fireRequest.tsx [hunk 8]
```diff
@@ -142,10 +142,11 @@ export const fireRequest = (
...
           if (error.response.status === 400 || error.response.status === 406) {
-            Notification.BadRequest({
-              errs: error.response.data,
-            });
-            return error.response;
+            console.log(error.response.data.detaillocal_ip_address);
+            // Notification.BadRequest({
+            //   errs: error.response.data,
+            // });
+            // return error.response;
```
Comment (rithviknishad) [inline src/Redux/fireRequest.tsx]: Why was this changed?

**Review by rithviknishad — CHANGES_REQUESTED** [review]
Only validation logic in this file needs to be modified: https://github.com/ohcnetwork/care_fe/blob/develop/src/Components/HCX/validators.ts

**Comment by syedfardeenjeelani** [comment]
> Only validation logic in this file needs to be modified: https://github.com/ohcnetwork/care_fe/blob/develop/src/Components/HCX/validators.ts

hey @rithviknishad  should it be something like this 
if (!value.policy_id.trim()) {
    return "some err message";
  } else if (!value.subscriber_id.trim()) {
    return "some err message";
  }
  if (enable_hcx) {
    if (!value.insurer_id?.trim()) {
      return "some err message";
    } else if (!value.insurer_name?.trim()) {
      return "some err message";
    }
  }  

also do we only have to change this particular file and nothing else

**Comment by rithviknishad** [comment]
> also do we only have to change this particular file and nothing else

yes

**Comment by nihal467** [comment]
@rithviknishad can you verify is it ready for testing

**Review by rithviknishad — CHANGES_REQUESTED** [review]
texts to be in i18n locale files

https://github.com/ohcnetwork/care_fe#translations

**Comment by syedfardeenjeelani** [comment]
> texts to be in i18n locale files
> 
> https://github.com/ohcnetwork/care_fe#translations

@rithviknishad  hello should i put them in common.json

**Comment by rithviknishad** [comment]
Put in `en.json`

**Comment by syedfardeenjeelani** [comment]
> Put in `en.json`

okay

**Comment by nihal467** [comment]
@syedfardeenjeelani is this PR ready for review

**Comment by syedfardeenjeelani** [comment]
@rithviknishad  i need some guidance i cant find en.json

**Comment by AdityaJ2305** [comment]
> @rithviknishad i need some guidance i cant find en.json

need to update the new changes in forked repo then pull

**Comment by rithviknishad** [comment]
yup

**Review by rithviknishad — APPROVED** [review]

**Comment by nihal467** [comment]
image
image

- [ ] Even though the validation is added, it's not blocking the user from submitting a patient registration form when the insurance fields are kept blank

**Comment by AdityaJ2305** [comment]
Hey @syedfardeenjeelani , Please check by adding this line `errors["insurance_details"] = insuranceDetailsError;` in `validateForm` func in `PatientRegister.tsx`

**Comment by rithviknishad** [comment]
@syedfardeenjeelani could you resolve the merge conflicts?

**Comment by rithviknishad** [comment]
ensure the en.json file has keys sorted. pre-commit should do it for you if setup properly.

**Comment by syedfardeenjeelani** [comment]
@nihal467  please test this

File: src/Locale/en.json
Comment (rithviknishad) [inline src/Locale/en.json:1]: why is the entire file having a change?

**Review by rithviknishad — CHANGES_REQUESTED** [review]

File: src/Locale/en.json
Comment (syedfardeenjeelani) [reply src/Locale/en.json:1]: pre-commit  was not sorting keys for me so i used https://codeshack.io/json-sorter/

Comment (rithviknishad) [reply src/Locale/en.json:1]: use `npm run sort-locales` instead; also why is pre-commit not sorting?

Comment (syedfardeenjeelani) [reply src/Locale/en.json:1]: i will look into it @rithviknishad  also can you please tell  me if i should   hard reset to commit [f5a265172e499e4f4016be1453751680b7eb416b](https://github.com/ohcnetwork/care_fe/pull/8681/commits/f5a265172e499e4f4016be1453751680b7eb416b) (fixed validation) or do something different from here

Comment (rithviknishad) [inline src/Locale/en.json:1]: Why is so many i18n keys deleted?

File: package-lock.json
Comment (rithviknishad) [inline package-lock.json:1]: Could you discard changes to the lock file?

**Review by rithviknishad — CHANGES_REQUESTED** [review]

**Comment by syedfardeenjeelani** [comment]
@rithviknishad  i will setup the repo again and raise a new pr

**Pull request closed without merging**